### PR TITLE
[SPARK-35341][PYTHON] Introduce BooleanExtensionOps

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?

- Introduce BooleanExtensionOps in order to make boolean operators `and` and `or` data-type-based.
- Improve error messages for operators `and` and `or`.

### Why are the changes needed?

Boolean operators __and__, __or__, __rand__, and __ror__ should be data-type-based

BooleanExtensionDtypes processes these boolean operators differently from bool, so BooleanExtensionOps is introduced.

These boolean operators themselves are also bitwise operators, which should be able to apply to other data types classes later. However, this is not the goal of this PR.

### Does this PR introduce _any_ user-facing change?

Yes. Error messages for operators `and` and `or` are improved.
Before:
```
>>> psser = ps.Series([1, "x", "y"], dtype="category")
>>> psser | True
Traceback (most recent call last):
...
pyspark.sql.utils.AnalysisException: cannot resolve '(`0` OR true)' due to data type mismatch: differing types in '(`0` OR true)' (tinyint and boolean).;
'Project [unresolvedalias(CASE WHEN (isnull(0#9) OR isnull((0#9 OR true))) THEN false ELSE (0#9 OR true) END, Some(org.apache.spark.sql.Column$$Lambda$1442/1725491640@6fb8afba))]
+- Project [__index_level_0__#8L, 0#9, monotonically_increasing_id() AS __natural_order__#12L]
   +- LogicalRDD [__index_level_0__#8L, 0#9], false

```

After:
```
>>> psser = ps.Series([1, "x", "y"], dtype="category")
>>> psser | True
Traceback (most recent call last):
...
TypeError: Bitwise or can not be applied to categoricals.
```

### How was this patch tested?

Unit tests.

Keyword: SPARK-35337

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

**Comment by xinrong-databricks** [comment]
py3.9 in Github actions is running pandas 1.2.4, which has a change in testing utils.
This PR is WIP for that.

**Comment by HyukjinKwon** [comment]
looks good from a cursory look

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py [hunk 1]
```diff
@@ -228,9 +230,272 @@ def test_rmod(self):
...
+        pser = pd.Series([True, False, None], dtype='bool')
+        psser = ps.from_pandas(pser)
+        self.assert_eq(True | pser, True | psser)
+        self.assert_eq(False | pser, False | psser)
+
+
+@unittest.skipIf(
+    not extension_dtypes_available, "pandas extension dtypes are not available"
+)
+class BooleanExtensionOpsTest(BooleanOpsTest, PandasOnSparkTestCase, TestCasesUtils):
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py:281]: Seems like we don't reuse tests from `BooleanOpsTest` at all. We should remove it for now?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py (hunk 1 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py]: Makes sense! Removed.

File: python/pyspark/pandas/data_type_ops/base.py [hunk 2]
```diff
@@ -81,6 +85,7 @@ class DataTypeOps(object, metaclass=ABCMeta):
     def __new__(cls, dtype: Dtype, spark_type: DataType):
         from pyspark.pandas.data_type_ops.binary_ops import BinaryOps
         from pyspark.pandas.data_type_ops.boolean_ops import BooleanOps
+        from pyspark.pandas.data_type_ops.boolean_ops import BooleanExtensionOps
```
Comment (pingsutw) [inline python/pyspark/pandas/data_type_ops/base.py:88]:
```suggestion
        from pyspark.pandas.data_type_ops.boolean_ops import BooleanOps, BooleanExtensionOps
```

Comment (pingsutw) [reply python/pyspark/pandas/data_type_ops/base.py:88]: We could combine these.

Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/base.py:88]: Good point! Combined.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**